src/Relation/Nullary/Finite/Setoid.agda

```diff
@@ -0,0 +1,148 @@
+import Relation.Binary.PropositionalEquality.Properties as ≡
+
+private
+  variable
+    c ℓ c′ ℓ′ : Level
+
+Ω : ∀ p → Setoid (lsuc p) p
+Ω p = record
+  { Carrier = Set p
+  ; _≈_ = λ P Q → (P → Q) × (Q → P)
+  ; isEquivalence = record
+    { refl = id , id
+    ; sym = swap
+    ; trans = λ (f , g) (f′ , g′) → f′ ∘ f , g ∘ g′
+    }
+  }
+
+Subset : Setoid c ℓ → (p : Level) → Set (c ⊔ ℓ ⊔ lsuc p)
+Subset X p = Func X (Ω p)
+
+FullSubset : {X : Setoid c ℓ} → Subset X 0ℓ
+FullSubset = record { to = λ _ → ⊤ }
+
+record EquivalenceRelation (X : Setoid c ℓ) r : Set (c ⊔ ℓ ⊔ lsuc r) where
+  infix 4 _∼_
+  open Setoid X
+  field
+    _∼_ : Rel Carrier r
+    ∼-resp-≈ : ∀ {x x′ y y′} → x ≈ x′ → y ≈ y′ → x ∼ y → x′ ∼ y′
+    ∼-isEquivalence : IsEquivalence _∼_
+
+  open IsEquivalence ∼-isEquivalence public renaming
+    ( refl                 to ∼-refl
+    ; trans                to ∼-trans
+    ; sym                  to ∼-sym
+    ; reflexive            to ∼-reflexive
+    ; isPartialEquivalence to ∼-isPartialEquivalence
+    )
+
+  ≈⇒∼ : ∀ {x y} → x ≈ y → x ∼ y
+  ≈⇒∼ q = ∼-resp-≈ refl q ∼-refl
+
+_⁻¹[_] : ∀ {p} {X : Setoid c ℓ} {Y : Setoid c′ ℓ′} →
+  Func X Y → Subset Y p → Setoid (c ⊔ p) ℓ
+_⁻¹[_] {X = X} f S =
+  On.setoid {B = Σ[ x ∈ X.Carrier ] S.to (f.to x)} X proj₁
+  where
+  module X = Setoid X
+  module f = Func f
+  module S = Func S
+
+_[_] : ∀ {p} {X : Setoid c ℓ} {Y : Setoid c′ ℓ′} →
+  Func X Y → Subset X p → Setoid (c ⊔ p) ℓ′
+_[_] {X = X} {Y} f S =
+  On.setoid {B = Σ[ x ∈ X.Carrier ] S.to x} Y (f.to ∘ proj₁)
+  where
+  module X = Setoid X
+  module f = Func f
+  module S = Func S
+
+↣⇒⊆ : ∀ {X : Setoid c ℓ} {Y : Setoid c′ ℓ′} → Injection Y X → Subset X (c′ ⊔ ℓ)
+↣⇒⊆ {X = X} {Y} m = record
+  { to = λ x → ∃ \ y → to y X.≈ x
+  ; cong = λ p →
+    (λ (y , q) → y , X.trans q p) , (λ (y , q) → y , X.trans q (X.sym p))
+  }
+  where
+  module X = Setoid X
+  module Y = Setoid Y
+  open Injection m
+
+_/_ : ∀ {r} (X : Setoid c ℓ) → EquivalenceRelation X r → Setoid c r
+X / R = record
+  { Carrier = Carrier
+  ; _≈_ = _∼_
+  ; isEquivalence = ∼-isEquivalence
+  }
+  where
+  open Setoid X
+  open EquivalenceRelation R
+
+include-/ : ∀ {r} {X : Setoid c ℓ} (R : EquivalenceRelation X r) →
+  Surjection X (X / R)
+include-/ {X = X} R = record
+  { to = id
+  ; cong = ≈⇒∼
+  ; surjective = λ y → y , ∼-refl
+  }
+  where
+  open Setoid X
+  open EquivalenceRelation R
+
+record IsStronglyFinite (X : Setoid c ℓ) (n : ℕ) : Set (c ⊔ ℓ) where
+  field
+    inv : Inverse X (≡.setoid (Fin n))
+
+record StronglyFinite (X : Setoid c ℓ) (n : ℕ) : Set (c ⊔ ℓ)  where
+  field
+    isFinite : IsStronglyFinite X n
+
+record Subfinite (X : Setoid c ℓ) : Set (c ⊔ ℓ) where
+  field
+    size : ℕ
+    inj : Injection X (≡.setoid (Fin size))
+
+record FinitelyEnumerable (X : Setoid c ℓ) : Set (c ⊔ ℓ) where
+  field
+    size : ℕ
+    srj : Surjection (≡.setoid (Fin size)) X
+
+record SubFinitelyEnumerable (X : Setoid c ℓ) c′ ℓ′
```

Comment:
The 'source literature' here is silly with its names, and this is copied here.

`InjectsIntoFinitelyEnumerable` ? And then `SurjectionFromFinitelyEnumerable` ?